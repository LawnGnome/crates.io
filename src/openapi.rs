use utoipa::OpenApi;
use utoipa_axum::router::OpenApiRouter;

#[derive(OpenApi)]
#[openapi(info(
    title = "crates.io",
    description = "API documentation for the [crates.io](https://crates.io/) package registry",
    terms_of_service = "https://crates.io/policies",
    contact(name = "the crates.io team", email = "help@crates.io"),
    license(),
    version = "0.0.0",
))]
pub struct BaseOpenApi;

impl BaseOpenApi {
    pub fn router<S>() -> OpenApiRouter<S>
    where
        S: Send + Sync + Clone + 'static,
    {
        OpenApiRouter::with_openapi(Self::openapi())
    }
}

#[cfg(test)]
mod tests {
    use crate::tests::util::{RequestHelper, TestApp};
    use http::StatusCode;
    use insta::assert_json_snapshot;

    #[tokio::test(flavor = "multi_thread")]
    async fn test_openapi_snapshot() {
        let (_app, anon) = TestApp::init().empty().await;

        let response = anon.get::<()>("/api/openapi.json").await;
        assert_eq!(response.status(), StatusCode::OK);
        assert_json_snapshot!(response.json());
    }
}
