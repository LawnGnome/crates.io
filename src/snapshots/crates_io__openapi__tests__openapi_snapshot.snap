---
source: src/openapi.rs
expression: response.json()
snapshot_kind: text
---
{
  "components": {},
  "info": {
    "contact": {
      "email": "help@crates.io",
      "name": "the crates.io team"
    },
    "description": "API documentation for the [crates.io](https://crates.io/) package registry",
    "license": {
      "name": ""
    },
    "termsOfService": "https://crates.io/policies",
    "title": "crates.io",
    "version": "0.0.0"
  },
  "openapi": "3.1.0",
  "paths": {
    "/api/v1/crates": {
      "get": {
        "description": "Called in a variety of scenarios in the front end, including:\n- Alphabetical listing of crates\n- List of crates under a specific owner\n- Listing a user's followed crates",
        "operationId": "crates_list",
        "responses": {
          "200": {
            "description": "Successful Response"
          }
        },
        "summary": "Returns a list of crates.",
        "tags": [
          "crates"
        ]
      }
    }
  },
  "servers": [
    {
      "url": "https://crates.io"
    },
    {
      "url": "https://staging.crates.io"
    }
  ]
}
